Python: rename internal parameters related to splicing and fix splicing

FRidh opened the pull request:
Follow-up to #104201, related to #105113.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

FRidh committed:
python.tests: use self.callPackage instead of super.callPackage

super was incorrectly possible until https://github.com/NixOS/nixpkgs/pull/104201
got merged.

FRidh committed:
Python: use pythonPackagesBuildHost instead of pythonForBuild

Follow-up to #104201, related to #105113.

Ericson2314 commented on pkgs/development/interpreters/python/cpython/2.7/default.nix:
```diff
@@ -38,6 +38,7 @@ assert x11Support -> tcl != null
+  buildPackages = pkgsBuildHost;
 
   pythonAttr = "python${sourceVersion.major}${sourceVersion.minor}";
   pythonForBuild = buildPackages.${pythonAttr};
```
like below
```suggestion
  inherit (passthru) pythonForBuild;
```

Ericson2314 commented on pkgs/development/interpreters/python/cpython/2.7/default.nix:
```diff
@@ -38,6 +38,7 @@ assert x11Support -> tcl != null
 with stdenv.lib;
 
 let
+  buildPackages = pkgsBuildHost;
```
Actually I wonder if we can get rid of this:

For example, `buildPackages.stdenv.cc` => `pkgsBuildBuild.targetPackages.stdenv.cc` which is kind of a mouthful, but I like it because it this is the `cc` for `BuildBuild` (holds for all `pkgsXxYy.targetPackages.stdenv.cc`, that is the `cc` for `XxYy`)

FRidh replied:
I use `buildPackages` here because this is what people are most familiar with.

Ericson2314 commented on pkgs/development/interpreters/python/default.nix:
```diff
@@ -99,7 +99,8 @@ with pkgs;
         pythonOlder = lib.versionOlder pythonVersion;
-        inherit hasDistutilsCxxPatch pythonForBuild;
+        inherit hasDistutilsCxxPatch;
+        pythonForBuild = pythonPackagesBuildHost;
```
I don't think this is right?

FRidh replied:
Why not? `passthruFun` is called with `pythonPackagesBuildHost = pkgsBuildHost.${pythonAttr};`.

Ericson2314 replied:
we had `pythonPackagesBuildHost == pythonForBuild.pkgs` before, i.e. `pythonForBuild` was a derivation

FRidh replied:
`pythonForBuild` is a derivation, and `pkgs` is the Python package set it comes with.

FRidh committed:
Python: rename parameters and arguments passed to passthru

As part of the splicing the build/host/target combinations of the interpreter
need to be passed around internally. The chosen names were not very clear,
implying they were package sets whereas actually there were derivations.

FRidh committed:
Python splicing: splice package sets, not Python derivation

In the original commit the various builds of Python were added to
`otherSplices`, instead of the intended Python package sets.

FRidh commented on pkgs/development/interpreters/python/cpython/default.nix:
```diff
@@ -67,11 +67,12 @@ let
+    pythonForBuildOnHost = pkgsBuildHost.${pythonAttr};
+    pythonForBuildOnTarget = pkgsBuildTarget.${pythonAttr};
+    pythonForHostOnHost = pkgsHostHost.${pythonAttr};
+    pythonForTargetOnTarget = pkgsTargetTarget.${pythonAttr} or {};
```
Maybe have it evaluate to `null` for TargetOnTarget`?

Ericson2314 commented on pkgs/development/interpreters/python/default.nix:
```diff
@@ -14,12 +14,12 @@ with pkgs;
+    , pythonOnBuildForTarget
+    , pythonOnHostForHost
+    , pythonOnTargetForTarget
+    , self # is pythonOnHostForTarget.pkgs
```
```suggestion
    , self # is pythonOnHostForTarget
```

FRidh committed:
Python: add todo note regarding pythonForBuild

Ericson2314 commented on pkgs/development/interpreters/python/default.nix:
```diff
@@ -99,7 +99,8 @@ with pkgs;
         pythonOlder = lib.versionOlder pythonVersion;
-        inherit hasDistutilsCxxPatch pythonForBuild;
+        inherit hasDistutilsCxxPatch;
+        pythonForBuild = pythonOnBuildForHost;
```
Maybe we add a comment here saying the `ForXx` mismatch is intentional and should be fixed by renaming `pythonForBuild`?
